From ec1c6f256e1d7f82abdcffac78f926f109e4df93 Mon Sep 17 00:00:00 2001
From: siashah <siashah@chromium.org>
Date: Thu, 08 Jun 2023 20:24:43 +0000
Subject: [PATCH] Do not show a snackbar if already showing one

This is a temporary fix and in the future we would overwrite the
dismiss the currently showing snackbar and show the new one with proper
garbage collection. crbug.com/1450942

(cherry picked from commit 721bef843b216bd9bcd57f671112c8a0f449a3ae)

Bug: 1450568
Change-Id: I064d76eccc06c6e23fcde9fd1122e87068ffa1c5
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4583539
Reviewed-by: Christoph Schwering <schwering@google.com>
Commit-Queue: Siddharth Shah <siashah@chromium.org>
Reviewed-by: Vinny Persky <vinnypersky@google.com>
Cr-Original-Commit-Position: refs/heads/main@{#1153332}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4601968
Cr-Commit-Position: refs/branch-heads/5735@{#1207}
Cr-Branched-From: 2f562e4ddbaf79a3f3cb338b4d1bd4398d49eb67-refs/heads/main@{#1135570}
---

--- a/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl.cc
+++ b/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl.cc
@@ -24,18 +24,15 @@
 }
 
 void AutofillSnackbarControllerImpl::Show() {
-  if (!autofill_snackbar_view_) {
-    autofill_snackbar_view_ = AutofillSnackbarView::Create(this);
+  if (autofill_snackbar_view_) {
+    // A snackbar is already showing. Ignore the new request.
+    return;
   }
+  autofill_snackbar_view_ = AutofillSnackbarView::Create(this);
   autofill_snackbar_view_->Show();
   base::UmaHistogramBoolean("Autofill.Snackbar.VirtualCard.Shown", true);
 }
 
-void AutofillSnackbarControllerImpl::SetViewForTesting(
-    AutofillSnackbarView* view) {
-  autofill_snackbar_view_ = view;
-}
-
 void AutofillSnackbarControllerImpl::OnActionClicked() {
   ManualFillingControllerImpl::GetOrCreate(GetWebContents())
       ->ShowAccessorySheetTab(autofill::AccessoryTabType::CREDIT_CARDS);
--- a/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl.h
+++ b/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl.h
@@ -26,7 +26,6 @@
 
   // Show the snackbar.
   void Show();
-  void SetViewForTesting(AutofillSnackbarView* view);
 
   // AutofillSnackbarController:
   void OnActionClicked() override;
--- a/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl_unittest.cc
+++ b/chrome/browser/ui/autofill/payments/autofill_snackbar_controller_impl_unittest.cc
@@ -20,13 +20,6 @@
 
 namespace autofill {
 
-class MockAutofillSnackbarView : public AutofillSnackbarView {
- public:
-  MockAutofillSnackbarView() = default;
-  void Show() override {}
-  void Dismiss() override {}
-};
-
 class AutofillSnackbarControllerImplTest
     : public ChromeRenderViewHostTestHarness {
  public:
@@ -34,7 +27,6 @@
 
   void SetUp() override {
     ChromeRenderViewHostTestHarness::SetUp();
-    controller()->SetViewForTesting(new MockAutofillSnackbarView());
     ManualFillingControllerImpl::CreateForWebContentsForTesting(
         web_contents(), mock_pwd_controller_.AsWeakPtr(),
         mock_address_controller_.AsWeakPtr(), mock_cc_controller_.AsWeakPtr(),
@@ -65,8 +57,6 @@
       "Autofill.Snackbar.VirtualCard.ActionClicked", 1, 0);
   controller()->OnDismissed();
 
-  // Reset the mock view.
-  controller()->SetViewForTesting(new MockAutofillSnackbarView());
   controller()->Show();
   controller()->OnActionClicked();
   // Verify that the count for both Shown and ActionClicked is incremented.
@@ -76,4 +66,22 @@
       "Autofill.Snackbar.VirtualCard.ActionClicked", 1, 1);
 }
 
+TEST_F(AutofillSnackbarControllerImplTest,
+       AttemptToShowDialogWhileAlreadyShowing) {
+  base::HistogramTester histogram_tester;
+  controller()->Show();
+  // Verify that the count for Shown is incremented and ActionClicked hasn't
+  // changed.
+  histogram_tester.ExpectUniqueSample("Autofill.Snackbar.VirtualCard.Shown", 1,
+                                      1);
+  histogram_tester.ExpectUniqueSample(
+      "Autofill.Snackbar.VirtualCard.ActionClicked", 1, 0);
+
+  // Attempt to show another dialog without dismissing the previous one.
+  controller()->Show();
+
+  // Verify that the count for Shown is not incremented.
+  histogram_tester.ExpectUniqueSample("Autofill.Snackbar.VirtualCard.Shown", 1,
+                                      1);
+}
 }  // namespace autofill
